This pull request enables new group widgets

Altonss opened the pull request:
This is just a draft because after a group widget is created, the group doesn't open up. Instead just the last window of the app opens. The group id is well transfered with the intent to the MainActivity, but then I don't really know how to do ^^ I'm sure it's a not a very difficult fix, but for the moment I'm stuck.

TheLastProject commented:
In MainActivity there is a note like this:

// Restore settings from Shared Preference

You probably need to edit that to be like: if this is opened from the group widget, use that group. Otherwise, restore from shared preference

Altonss commented:
Ah ok I understand. But I don't really know how to write the "if this is opened from the group widget" logic... I'm new to android development :)

Altonss commented:
It works now on my device! Feel free to make a review or propose some changes. As I'm new to android development, don't hesitate to make suggestions/corrections/bug reports !

Altonss commented:
I noticed a bug: when changing a group name after a widget has been created for this group, the widget doesn't work anymore.
At least like for the cards, the widget should continue to work even if there the widget doesn't update the icon and the text.

Altonss commented:
I just added the following: an error message now appears if the group name has changed or has been deleted. That's the best thing I could do for this case :) I don't see other errors, so for me this PR is fully functional :)

TheLastProject commented:
Please don't edit the MainActivity to also list groups, it complicates the code a lot and makes it hard to maintain.

You code is honestly overly complex and given how simple selecting a group is, you could use Android's built-in `android.R.layout.simple_list_item_2` for choosing a group, then you don't even need to create or edit any layout. Something like this: https://stackoverflow.com/a/18529511

Altonss commented:
> Please don't edit the MainActivity to also list groups, it complicates the code a lot and makes it hard to maintain.

Well I didn't modify MainActivity to also list groups... at least I don't think so. I just implemented a way to check if a certain group has been passed by the intent, and to open this group up. But indeed I can simplify this part as some parts are written twice!


> You code is honestly overly complex and given how simple selecting a group is, you could use Android's built-in `android.R.layout.simple_list_item_2` for choosing a group, then you don't even need to create or edit any layout. Something like this: https://stackoverflow.com/a/18529511

Oh I'm sorry if it is too complicated :/ I just went the "simple" route and took already created parts in the code. I kind of reused the ManageGroups Layout, that looked good to me, but I can change the layout to the one you proposed.

So to resume the changes I need to make: refactor/simplify the "detect intent group" test in MainActivity and use the standart layout instead of the one I created?

Altonss commented:
I simplified MainActivity.

For the group selection code complexity,
> You code is honestly overly complex and given how simple selecting a group is, you could use Android's built-in `android.R.layout.simple_list_item_2` for choosing a group, then you don't even need to create or edit any layout. Something like this: https://stackoverflow.com/a/18529511


I don't really know if I should change something. It looks pretty good right now in my opinion and is pretty consistent with the group management view. I just replaced the 4 buttons (up, down, edit and delete) with a select button.
If you really want me to change the code, that would be feasible. But using the suggested ArrayAdapter seems not to be a drop-in replacement, because for the moment all is coded for a RecyclerView...

TheLastProject commented on app/src/main/java/protect/card_locker/GroupShortcutConfigure.java:
```diff
@@ -0,0 +1,96 @@
+        // aback button is pressed.
+        setResult(RESULT_CANCELED);
+
+        setContentView(R.layout.main_activity);
```
I created a new clean base layout for this, please use `R.layout.simple_toolbar_list_activity` instead (see the changes to `CardShortcutConfigure.java`). Then you also don't have to hide the FAB and mess around with the LayoutManager and all

TheLastProject commented on app/src/main/java/protect/card_locker/ShortcutHelper.java:
```diff
@@ -155,4 +155,26 @@ Bitmap createAdaptiveBitmap(@NotNull Bitmap in, int paddingColor) {
+        intent.setFlags(intent.getFlags() | Intent.FLAG_ACTIVITY_SINGLE_TOP);
+        final Bundle bundle = new Bundle();
+        // Done like with the cards but with group._id (a string)
+        bundle.putString("id", group._id);
```
Please use a more descriptive ID and make it a static final String in MainActivity to prevent typo risks. `id` is really the type of value that may be accidentally used for something else than the group ID one day.

Altonss replied:
With the current syntax, I cannot set this string to be static in MainActivity :/ I would need to find out what to change to be able to make it a static variable

TheLastProject commented on app/src/main/java/protect/card_locker/ShortcutHelper.java:
```diff
@@ -155,4 +155,26 @@ Bitmap createAdaptiveBitmap(@NotNull Bitmap in, int paddingColor) {
+        final Bundle bundle = new Bundle();
+        // Done like with the cards but with group._id (a string)
+        bundle.putString("id", group._id);
+        bundle.putBoolean("view", true);
```
This boolean seems unused?

Altonss replied:
Yes I think you are right. I noticed it might also be the case with "createShortcutBuilder" (for the cards shortcuts) just above it?

TheLastProject commented on app/src/main/java/protect/card_locker/ShortcutHelper.java:
```diff
@@ -155,4 +155,26 @@ Bitmap createAdaptiveBitmap(@NotNull Bitmap in, int paddingColor) {
+                .setShortLabel(group._id)
+                .setLongLabel(group._id)
+                .setIntent(intent)
+                .setIcon(icon);
```
Given groups don't inherently have their own icon, why not just use Catima's logo here instead? That way it is also clear from the widget that this won't open a card.

Altonss replied:
That is a good idea, however how could we differentiate between a group widget and the app icon except the title? Would that be enough?

TheLastProject replied:
I think the title would be sufficient for now

TheLastProject commented on app/src/main/java/protect/card_locker/ShortcutHelper.java:
```diff
@@ -155,4 +155,26 @@ Bitmap createAdaptiveBitmap(@NotNull Bitmap in, int paddingColor) {
+
+        IconCompat icon = IconCompat.createWithAdaptiveBitmap(iconBitmap);
+
+        return new ShortcutInfoCompat.Builder(context, group._id)
```
> I just added the following: an error message now appears if the group name has changed or has been deleted. That's the best thing I could do for this case :)

You should implement an [updateShortcuts](https://developer.android.com/reference/androidx/core/content/pm/ShortcutManagerCompat#updateShortcuts(android.content.Context,%20java.util.List%3Candroidx.core.content.pm.ShortcutInfoCompat%3E)) function for groups. I believe it uses the group name here (`group._id`). But it would be nicer if you would prefix the shortcut names to make them clearly different from loyalty cards (like `GROUP_groupname`). You will have to call that function whenever a group is edited or deleted.

TheLastProject replied:
That's not what I meant. I meant it like this:

If group._id = Fashion, don't use "Fashion" as shortcut ID, but use "GROUP_Fashion". That way you can tell which shortcuts are group-related in the code easily because they all start with "GROUP_". That makes searching for the right shortcut in the updateShortcuts function way easier (no possibility to get confused with cards).

Altonss replied:
I see what you mean. Indeed I already tried to implement a updateShortcuts and removeShortchuts functions for group shortcuts, but nothing was happening with my change: the name of the shortcut doesn't change and the shortcut doesn't get deleted. This is also the same for card shortcuts on my android 5.0.2 device, could this issue be related to a too low api level?

TheLastProject replied:
I do get comment notifications, sorry, been a bit busy.

>Does the updateShortcut and removeShortcut work for you with cards, e.g. is the shortcut title and icon updated/removed if the card is updated/removed?

updating works, but removing doesn't. I've done this:
1. Create a shortcut for a card named "Clothing store"
2. Rename the card to "test" -> shortcut changed
3. Changed the card image -> shortcut changed
4. Removed the shortcut -> no change

I'll make a new GitHub issue for that, the deletion should do something. Created #686 for that.

Altonss replied:
> I do get comment notifications, sorry, been a bit busy.

That's normal, no problem :)

> * Create a shortcut for a card named "Clothing store"
> 
> * Rename the card to "test" -> shortcut changed
> 
> * Changed the card image -> shortcut changed

This doesn't work on Android 5.0.2 device, so I will open an issue for that (probably related to the old api level). I can try to implement updateShortcut for group shortcuts, but I will not be able to test it on my device :/ So should I implement this in this PR?

TheLastProject replied:
I think I first want to see if we can fix this not working for cards on Android 5, if that can be fixed it could be implemented here too. Otherwise I will review it without that feature and add it myself for newer Android versions.

TheLastProject reviewed (changes requested):
>and is pretty consistent with the group management view

The main problem is that the group management view actually looks pretty bad, so basing it on that doesn't really sound like much of an improvement. But I will accept it I guess. I also didn't mean to change it to an ArrayAdapter, I may just make that change myself later.

Altonss commented:
> The main problem is that the group management view actually looks pretty bad, so basing it on that doesn't really sound like much of an improvement. But I will accept it I guess. I also didn't mean to change it to an ArrayAdapter, I may just make that change myself later.

Ah I didn't know that you consider the group management to look bad. I don't think it looks that bad, but I understand that you might have some ideas for improvements :) Anyway I'm sorry if my implementation isn't the best and I will look at the changes you requested.

Altonss commented:
Apart from the icon for the group shortcut and the updateShortcut discussions, I have implemented all requested changes as far as I could.
But while testing, I noticed a bug: when opening a group widget, then tapping on the home button, then open a card widget and then open a group widget again, the view stays at the card and the group doesn't open :/

Altonss commented:
> But while testing, I noticed a bug: when opening a group widget, then tapping on the home button, then open a card widget and then open a group widget again, the view stays at the card and the group doesn't open :/
> EDIT: After further testing, I noticed that this bug is not really related to my implementation of group widgets. It also happens when doing card widget > group widget > card widget or other combinations also with the app icon itself...

@TheLastProject Should I open a separate issue for that? Can this group widget branch be merged even if there is this bug or should we fix the bug before to avoid user confusion?

TheLastProject commented:
> > But while testing, I noticed a bug: when opening a group widget, then tapping on the home button, then open a card widget and then open a group widget again, the view stays at the card and the group doesn't open :/
> > EDIT: After further testing, I noticed that this bug is not really related to my implementation of group widgets. It also happens when doing card widget > group widget > card widget or other combinations also with the app icon itself...
> 
> @TheLastProject Should I open a separate issue for that? Can this group widget branch be merged even if there is this bug or should we fix the bug before to avoid user confusion?

I'm not sure how that bug isn't inside of this PR? If you can reproduce something on the master branch please create a new GitHub issue.

Altonss commented:
> I'm not sure how that bug isn't inside of this PR? If you can reproduce something on the master branch please create a new GitHub issue.

I cannot reproduce it with the last stable release. It's a specific issue related to group widget, card widget and the app launch icon. Something is working great and it may also include the implementation of the card widget (the widget in general). I will try to look at this issue. What needs to be done: When we click on a shortcut, it should open it whatever has been opened before

Altonss commented:
I just pushed a fix for the interference of the card and group widgets...but I don't understand why it doesn't build anymore. It works on my machine :(

TheLastProject commented:
> I just pushed a fix for the interference of the card and group widgets...but I don't understand why it doesn't build anymore. It works on my machine :(

Hmm, I think GitHub tries to applies your changes on top of master, which has a lot of DB-related changes (see #680). If you pull the master branch into your branch you will see the same build errors now.

Altonss commented:
> Hmm, I think GitHub tries to applies your changes on top of master, which has a lot of DB-related changes (see #680). If you pull the master branch into your branch you will see the same build errors now.

Ah I didn't know this behavior, weird as I didn't merge the latest master branch in this PR. So should I import the current master in my branch and see where I should fix it?

TheLastProject commented:
> > Hmm, I think GitHub tries to applies your changes on top of master, which has a lot of DB-related changes (see #680). If you pull the master branch into your branch you will see the same build errors now.
> 
> Ah I didn't know this behavior, weird as I didn't merge the latest master branch in this PR. So should I import the current master in my branch and see where I should fix it?

Yup exactly!

Basically, all the database stuff changed so your `db.function(value)` like things should become `DBHelper.function(db, value)`. It's a fairly easy change once you figure it out :)

Altonss commented:
Well I am testing out the latest changes, but with the merge of the new DB changes, I'm not able to create any shortcut anymore. The app crashes before even entering the selector. Is that maybe due to 2 errors in DBHelper? (in Android Studio 2 errors are detected): line 111 and line 266.

TheLastProject commented:
The `tokenize=unicode61` stuff is not an error, it's just Android Studio not understanding that valid syntax.

If you get a crash, well, what is the crash exactly? It should print a stacktrace while running the app through Android Studio.

Altonss commented:
> If you get a crash, well, what is the crash exactly? It should print a stacktrace while running the app through Android Studio.

I'm running it on my own device, for some reason didn't manage to start the emulators, or they were too slow.

TheLastProject commented:
> I'm running it on my own device, for some reason didn't manage to start the emulators, or they were too slow.

That shouldn't matter, as long as you use adb you will still just see the stack trace in Android Studio when the app crashes no matter if it's running on your device or in an emulator.

The only time you won't see the crash log in Android Studio is if you don't run the app through Android Studio but instead build it and manually install the .apk. While I really wouldn't recommend going that route as it is extremely slow and tedious in comparison, [Scoop](https://f-droid.org/en/packages/taco.scoop/) can help you in that case.

Altonss commented:
Here is the crash I get:
`
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: me.hackerchick.catima.debug, PID: 13120
    java.lang.RuntimeException: Unable to instantiate activity ComponentInfo{me.hackerchick.catima.debug/protect.card_locker.CardShortcutConfigure}: java.lang.NullPointerException: Attempt to invoke virtual method 'android.database.sqlite.SQLiteDatabase android.content.Context.openOrCreateDatabase(java.lang.String, int, android.database.sqlite.SQLiteDatabase$CursorFactory, android.database.DatabaseErrorHandler)' on a null object reference
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2335)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2544)
        at android.app.ActivityThread.access$900(ActivityThread.java:163)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1367)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5608)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1397)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1192)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.database.sqlite.SQLiteDatabase android.content.Context.openOrCreateDatabase(java.lang.String, int, android.database.sqlite.SQLiteDatabase$CursorFactory, android.database.DatabaseErrorHandler)' on a null object reference
        at android.content.ContextWrapper.openOrCreateDatabase(ContextWrapper.java:274)
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:223)
        at android.database.sqlite.SQLiteOpenHelper.getReadableDatabase(SQLiteOpenHelper.java:187)
        at protect.card_locker.CardShortcutConfigure.<init>(CardShortcutConfigure.java:21)
        at java.lang.reflect.Constructor.newInstance(Native Method)
        at java.lang.Class.newInstance(Class.java:1656)
        at android.app.Instrumentation.newActivity(Instrumentation.java:1071)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2325)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2544) 
        at android.app.ActivityThread.access$900(ActivityThread.java:163) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1367) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:135) 
        at android.app.ActivityThread.main(ActivityThread.java:5608) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1397) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1192) 
`

TheLastProject commented:
```
android.database.sqlite.SQLiteOpenHelper.getReadableDatabase(SQLiteOpenHelper.java:187) at protect.card_locker.CardShortcutConfigure.<init>(CardShortcutConfigure.java:21)
```

That's the important part, so CardShortcutConfigure.java, line 21. It doesn't look like you changed anything there so maybe you found a crash that's in the master branch? What did you do to reproduce it? I can't reproduce the crash you mentioned on the master branch. I haven't tried your branch yet for this, it's been a busy week sorry.

Altonss commented:
> ```
> android.database.sqlite.SQLiteOpenHelper.getReadableDatabase(SQLiteOpenHelper.java:187) at protect.card_locker.CardShortcutConfigure.<init>(CardShortcutConfigure.java:21)
> ```
> 
> That's the important part, so CardShortcutConfigure.java, line 21. It doesn't look like you changed anything there so maybe you found a crash that's in the master branch? What did you do to reproduce it? I can't reproduce the crash you mentioned on the master branch. I haven't tried your branch yet for this, it's been a busy week sorry.

No worries, you can take your time :) I don't think I changed anything in this place compared to master. Will try the master branch on the same device and maybe open an issue if I encounter the problem again. To reproduce the issue on my branch: just try to create a shortcut.

Altonss commented:
It's the same in the main branch, just created an issue.

Altonss commented:
I'm sorry that this takes so much time for me to fix. Currently I don't have much time, but I plan to resume the work on this PR soon:copyright: :laughing: 
But I anyone has an idea of how to help for the update/delete shortcut functionnality it would be great. I tried several things but it was not working :/
Thanks!

TheLastProject commented:
No worries. I could maybe do the update/delete part myself if you can't run an emulator with a less broken Android version but no promises because time and stuff

TheLastProject commented:
Looking at the "Resolve conflicts" button on GitHub, the conflicts are really small, should be easy to resolve in the web UI.

But you can also use the command line, look at the text "Use the [web editor](https://github.com/CatimaLoyalty/Android/pull/669/conflicts) or the to resolve conflicts." GitHub shows on this PR

Altonss commented:
Yes I now I could use this, but the group widget selector was based on the old UI and has therefor not been adapted to the new UI...

TheLastProject commented:
Hmm, sorry, no suggestion there :(

Altonss commented:
> Hmm, sorry, no suggestion there :(

All right, no problem :) Just wanted to ask in case there was a way!

Altonss commented:
This PR now works again ;) But the issue with the group shortcut update isn't resolved and I'm really stuck. I think the logic for card updates is in ShortcutHelper, but I don't really understand it...

ping @TheLastProject, do you know how I could do this?

TheLastProject commented:
Yeah, I think it is in ShortcutHelper too but I don't know for sure either. That code is still from Loyalty Card Keychain and I never touched it because it always worked for me.

Have you tried running an emulator of a newer Android version than Android 5 in Android Studio yet? Because the last testing we did did seem to show that it's an Android 5 issue also with existing shortcuts.

I think something went wrong with merging the new code into this PR by the way, I see a lot of changes in it unrelated to new group widgets :(

Altonss commented:
> Yeah, I think it is in ShortcutHelper too but I don't know for sure either. That code is still from Loyalty Card Keychain and I never touched it because it always worked for me.

The code seems pretty "complicated" for me ^^

> Have you tried running an emulator of a newer Android version than Android 5 in Android Studio yet? Because the last testing we did did seem to show that it's an Android 5 issue also with existing shortcuts.

I just tested on Android 13 ;) What doesn't work:
- [ ] delete a group doesn't delete a shortcut, bug already happening for cards #686 
- [ ] update group nape doesn't update shortcut, but this works with cards (not on Android 5 though see #687)

> I think something went wrong with merging the new code into this PR by the way, I see a lot of changes in it unrelated to new group widgets :(

Really? I just checked again and it seems to me that all changes are related to group shortcuts (except a few blank lines here and there that I could clean up (maybe they were added by the ide?).

Altonss commented:
@TheLastProject After a bit of thinking, I have an idea why the updating might not work:
For cards, the id always remains the same, no? Because if this is the case, then maybe it doesn't work for groups because the id changes it the name of the group changes...

TheLastProject commented:
> For cards, the id always remains the same, no? Because if this is the case, then maybe it doesn't work for groups because the id changes it the name of the group changes...

I completely forgot I coded it that way! Ouch, that's not great.

Okay, so, we'll first have to (and let's do this in another PR):
1. Edit the `LoyaltyCardDbGroups` table to contain a new ID column that's an integer.
2. Edit the `LoyaltyCardDbIdsGroups` table to refer to the group ID instead of group name.
3. Migrate all current memberships to the new setup.

I think the easiest way is to just create 2 temporary tables and then create the new stuff like I did at newVersion 10 in the DBHelper class.

It's going to be a bit of a pain but a requirement to make this work well I'm afraid :(

Altonss commented:
> > For cards, the id always remains the same, no? Because if this is the case, then maybe it doesn't work for groups because the id changes it the name of the group changes...
> 
> I completely forgot I coded it that way! Ouch, that's not great.

So think that this issue might be the cause of the widget not updating because the id changes?

> Okay, so, we'll first have to (and let's do this in another PR):
> 
>     1. Edit the `LoyaltyCardDbGroups` table to contain a new ID column that's an integer.
> 
>     2. Edit the `LoyaltyCardDbIdsGroups` table to refer to the group ID instead of group name.
> 
>     3. Migrate all current memberships to the new setup.
> 
> 
> I think the easiest way is to just create 2 temporary tables and then create the new stuff like I did at newVersion 10 in the DBHelper class.
> 
> It's going to be a bit of a pain but a requirement to make this work well I'm afraid :(

Maybe I can open a new issue to track those 3 points :)

TheLastProject commented:
> So think that this issue might be the cause of the widget not updating because the id changes?

Yeah, I'm pretty sure that's the reason. Nice find :)

> Maybe I can open a new issue to track those 3 points :)

Sounds great :)
